### PR TITLE
air bitwise operation bug poc test

### PR DESCRIPTION
**Pull request by 0xKanekiKen**
Addressing part 1 of #206. 

- [x] Add a failing test to the [bitwise_pow2 unit tests](https://github.com/maticnetwork/miden/blob/next/air/src/aux_table/bitwise_pow2/tests.rs) to demonstrate the issue. This can be done by manually creating a test frame with different selectors, testing against a row that is not the first row in the cycle, and asserting that enforce constraints should result in an array that is not all zeros. This can be done first as one PR.
- [ ] Define new constraints on the selectors so that they can only change at the first row of a new periodic cycle, then update the mdbook docs for the Auxiliary Table.
- [ ] Implement the new constraints in the [Bitwise & Power of Two co-processor](https://github.com/maticnetwork/miden/blob/next/air/src/aux_table/bitwise_pow2/mod.rs)

### CODE REVIEW

**Review by grjte — CHANGES_REQUESTED** [review]
This is a good start, thank you! However, I think we should actually take a simpler approach for these failure tests and define the input values more explicitly, rather than writing code to generate them. This means they will be written differently from the existing tests. 

There are a few reasons for this:

1. We typically only use randomized inputs when we’re testing validity. For a test of a failure case, I think it’s sufficient to pick one representative case and run a single unit test, rather than using a proptest. This will be enough to demonstrate and test each bug or edge case without slowing down the test suite too much. In this case, I think the randomization of the rows doesn’t actually gain us anything in terms of extra test coverage, since the aggregation of bits over a single row is always the same and the periodic values are always `[Felt::ZERO, Felt::ONE]`. Effectively all that’s happening is we’re testing different shifts of the `a` and `b` values, and there’s no need to do that if we just pick 2 good values that will demonstrate the failure for each set of selectors.
2. When the input is generated, it’s harder to see what’s happening. Ideally, it should be really explicitly clear what’s being tested in these failure tests so that any new person looking at this test will be able to immediately understand the issue outlined in #206.
3. I think code that generates input for failure cases is more likely to have bugs or be overly complex (and therefore harder to reason about). For example, I think there’s actually an issue with how you’re constructing the frames. Here are the bitwise representations of your inputs: 1111_1111_1111_1111_1111_1111_1111_1011, 1011_1110_0000_0010_1000_0100_0101_1010. In the third-highest-order limb you can see that one is all 1’s and the other is all 0’s. That means there should be a failure when calling XOR and OR for one of the rows, but there currently isn’t.


For these reasons, my preference would be to use unit tests that define the current row, next row, and periodic values explicitly. This could look something like the following, and some of it could still be abstracted into a helper function.

```
use vm_core::{
    bitwise::TRACE_WIDTH as BITWISE_TRACE_WIDTH,
    TRACE_WIDTH,
};

#[test]
fn test_example() {
    let mut current = vec![Felt::ZERO; TRACE_WIDTH];
    let mut next = vec![Felt::ZERO; TRACE_WIDTH];
    let current_bitwise = vec![ …define 14 values explicitly for the current row… ];
    let next_bitwise = vec![ …define 14 values explicitly for the next row… ];
    current[BITWISE_TRACE_OFFSET..BITWISE_TRACE_OFFSET + BITWISE_TRACE_WIDTH]
        .copy_from_slice(&current_bitwise);
    next[BITWISE_TRACE_OFFSET..BITWISE_TRACE_OFFSET + BITWISE_TRACE_WIDTH]
        .copy_from_slice(&next_bitwise);

    let frame = EvaluationFrame::from_rows(current, next);
    let periodic_values = [Felt::ZERO, Felt::ONE];
    let mut result = [Felt::ZERO; bitwise::NUM_CONSTRAINTS];
    let expected = result;

    bitwise::enforce_constraints(&frame, &periodic_values, &mut result, Felt::ONE);

    assert_!(expected, result);
}
```


Finally, a couple of general comments:
1. Since you’re adding the first “failure” tests there may be opportunities to abstract functionality that’s shared with the existing “success” tests, which means some refactoring of the existing code to avoid duplication. Please feel free to do this refactoring as required - this goes for any future PRs too :)
2. Again, since these are the first “failure” tests there isn’t a framework in place for this yet, but we’ll want to add quite a few of these, so it’s fine to make functions that are fully generalized & could be used later for other failure tests (an example would be a function that takes frame & periodic_values as input). For anything that’s not fully generalized to work for any generic failure test, it should be named with something descriptive related to the specific thing you’re testing (in this case the selectors).

Please let me know if you have any questions or thoughts on this!

File: air/src/aux_table/bitwise_pow2/bitwise.rs [hunk 1]
```diff
@@ -443,3 +443,84 @@ pub fn get_test_frame(
...
+        next[A_COL_RANGE.start + idx] = Felt::new((next_a >> idx) & 1);
+        next[B_COL_RANGE.start + idx] = Felt::new((next_b >> idx) & 1);
+    }
+
+    // Set the output.
+    let output_curr = get_output_on_bitwise_operation(a, b, operation_1);
+    let output_next = get_output_on_bitwise_operation(a, b, operation_2);
+
+    current[OUTPUT_COL_IDX] = Felt::new((output_curr >> current_shift) as u64);
+    next[OUTPUT_COL_IDX] = Felt::new((output_next >> next_shift) as u64);
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/bitwise.rs]: makes sense now. I overlooked this part earlier. Thanks, for the explanation. I've removed this part in the latest push.

**Review by grjte — CHANGES_REQUESTED** [review]
Thank you! I like this approach - the main test is very explicit and readable now. I left comments inline, but there are 2 general themes:

1. "magic numbers": none of our constructions are set in stone, so I think it's important to use predefined constants for indexing rather than integers so that if something changes we can update constants in one place rather than having to search through tests for hard-coded values.
2. cycle vs row: There seems to be confusion about how the periodic cycles work, which is helpful to know, because it means we should add more information about this to the mdbook docs. I'll leave an explanation below, and please ask me questions if it's unclear, because I'll probably update the  docs with something similar to this.

## Periodic cycles
A `periodic cycle` is the number of rows X before a specified set of values in a `periodic column` will repeat. These values will repeat every X rows for the entire length of the trace. We use periodic cycles for operations that require more than one row, because it gives us a way to use selectors that specify the rows where different constraints need to be applied during a particular operation. 

Each processor can have different periodic cycles with one or more periodic columns. When evaluating transition constraints, we have access to all of the periodic values for the current row in the frame. This means that if there are 10 periodic columns then we’ll have an array of 10 values containing the periodic value of each column at the current row.

For example, a processor that requires 8 rows to perform an operation, such as the bitwise processor, will need to apply different constraints at the first row of the operation than at the last row. Periodic columns give us access to selectors that help us apply these constraints correctly. If the periodic cycle is length 8 and there are 2 periodic columns, then at trace row 9 we will have access to two periodic values, and they will be the same as the periodic values that were available at row 1. This makes it easy to apply the constraints that should happen at row 1 of the operation.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 2]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+    (result, expected)
+}
+
+/// Aggregates the decomposed value of input bits.
+fn aggregate_input_bits(bits_array: &[Felt], prev_input: Felt) -> Felt {
+    bitwise::agg_bits(bits_array, 0) + prev_input * Felt::new(2_u64.pow(4))
+}
+
+/// Generates the output value from decomposed input using a specified input operation (AND, OR, XOR)
+fn bitwise_operation_on_decomposed_input(
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Done. Updated to `aggregate_row_output`

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 3]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+    next: &[Felt],
+    operation1: Selectors,
+    operation2: Selectors,
+) -> ([Felt; 8], [Felt; 8]) {
+    let cycle = [Felt::ZERO, Felt::ONE];
+    let mut current_cycle = vec![Felt::ZERO; TRACE_WIDTH];
+    let mut next_cycle = vec![Felt::ZERO; TRACE_WIDTH];
+
+    let mut current_bitstate = vec![Felt::ZERO; 14];
+    let mut next_bitstate = vec![Felt::ZERO; 14];
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Updated the code.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 4]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+    // Hardcoded previous column a,b and output values.
+    let prev_input_a = Felt::new(10);
+    let prev_input_b = Felt::new(9);
+    let prev_output = Felt::new(8);
+
+    // calculating the input values for current and next cycles.
+    let input_curr_a = aggregate_input_bits(&curr[0..4], prev_input_a);
+    let input_curr_b = aggregate_input_bits(&curr[4..8], prev_input_b);
+    let input_next_a = aggregate_input_bits(&next[0..4], input_curr_a);
+    let input_next_b = aggregate_input_bits(&next[4..8], input_curr_b);
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Yep, now using `NUM_DECOMP_BITS`.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 5]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+
+    // calculating the ouptut values for current and next cycles.
+    let output_agg_curr_op1 = bitwise_operation_on_decomposed_input(&curr, operation1, prev_output);
+    let output_agg_next_op1 =
+        bitwise_operation_on_decomposed_input(&next, operation2, output_agg_curr_op1);
+
+    // First two columns would be selector.
+    // First & second cycle has been assigned to operation2 and operation1 respectively
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Thanks for explaining. I've updated the comments and variables name accordingly.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 6]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+    next_bitstate[3] = input_next_b;
+
+    for idx in 0..NUM_BITS_PER_CYCLE * 2 {
+        current_bitstate[4 + idx] = curr[idx];
+        next_bitstate[4 + idx] = next[idx];
+    }
+
+    current_bitstate[NUM_SELECTORS + NUM_SELECTORS + NUM_BITS_PER_CYCLE * 2 - 1] =
+        output_agg_curr_op1;
+    next_bitstate[NUM_SELECTORS + NUM_SELECTORS + NUM_BITS_PER_CYCLE * 2 - 1] = output_agg_next_op1;
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Done

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 7]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+
+    current_bitstate[2] = input_curr_a;
+    current_bitstate[3] = input_curr_b;
+    next_bitstate[2] = input_next_a;
+    next_bitstate[3] = input_next_b;
+
+    for idx in 0..NUM_BITS_PER_CYCLE * 2 {
+        current_bitstate[4 + idx] = curr[idx];
+        next_bitstate[4 + idx] = next[idx];
+    }
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: 1. Done
2. Does we get a performance boost or it's the rust way?

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 8]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+/// Contraints are applied to this frame which outputs it's result along with the expected result.  
+fn test_bitwise_with_customized_frame(
+    curr: &[Felt],
+    next: &[Felt],
+    operation1: Selectors,
+    operation2: Selectors,
+) -> ([Felt; 8], [Felt; 8]) {
+    let cycle = [Felt::ZERO, Felt::ONE];
+    let mut current_cycle = vec![Felt::ZERO; TRACE_WIDTH];
+    let mut next_cycle = vec![Felt::ZERO; TRACE_WIDTH];
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Udpated the variable name accordingly. Thanks, again for explaining.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 9]
```diff
@@ -1,12 +1,49 @@
...
+    TRACE_WIDTH,
+};
 
 use proptest::prelude::*;
 
 // UNIT TESTS
 // ================================================================================================
 
+#[test]
+fn test_bitwise_operation_fail() {
```
[the remark this answers is not in the document]
Comment (0xKanekiKen) [reply air/src/aux_table/bitwise_pow2/tests.rs]: Done

**Comment by 0xKanekiKen** [comment]
@grjte: I've done some more changes to the code. Let me know if it's fine or not. Thanks, again, for explaining periodic cycle in such detail. Whatever confusion I'd earlier on it has been cleared.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 10]
```diff
@@ -55,6 +92,112 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+    }
+
+    current_bitstate[NUM_SELECTORS + NUM_SELECTORS + NUM_BITS_PER_CYCLE * 2 - 1] =
+        output_agg_curr_op1;
+    next_bitstate[NUM_SELECTORS + NUM_SELECTORS + NUM_BITS_PER_CYCLE * 2 - 1] = output_agg_next_op1;
+
+    current_cycle[BITWISE_TRACE_OFFSET..BITWISE_TRACE_OFFSET + BITWISE_TRACE_WIDTH]
+        .copy_from_slice(&current_bitstate);
+    next_cycle[BITWISE_TRACE_OFFSET..BITWISE_TRACE_OFFSET + BITWISE_TRACE_WIDTH]
+        .copy_from_slice(&next_bitstate);
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/aux_table/bitwise_pow2/tests.rs]: I would probably define a range like `BITWISE_TRACE_RANGE` and then index it like `next_cycle[BITWISE_TRACE_RANGE]`

File: air/src/aux_table/bitwise_pow2/tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/aux_table/bitwise_pow2/tests.rs]: It should also be faster because it uses memcpy

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 11]
```diff
@@ -1,12 +1,68 @@
...
+        BITWISE_XOR,
+        BITWISE_AND,
+    );
+    let (result, expected) = get_bitwise_frame_result(frame, cycle);
+    assert_ne!(result, expected);
+
+    let frame =
+        get_test_frame_with_two_selectors(&current_bitwise, &next_bitwise, BITWISE_OR, BITWISE_XOR);
+    let (result, expected) = get_bitwise_frame_result(frame, cycle);
+    assert_ne!(result, expected);
```
Comment (grjte) [inline air/src/aux_table/bitwise_pow2/tests.rs:63]: With `assert_ne` at this level, I would prefer to define `expected` directly in this function and only return `result` from `get_bitwise_frame_result`. Since it's returning the same `expected` value (zeros) every time and that value doesn't depend on anything in the function, there's no need to have it come the function, and it could be defined just one time at this level.

**Review by grjte — CHANGES_REQUESTED** [review]
Looks great! I just left 2 final things inline, then it'll be ready to merge.

File: air/src/aux_table/bitwise_pow2/tests.rs [hunk 12]
```diff
@@ -55,6 +126,101 @@ fn test_bitwise_frame(operation: Selectors, a: u32, b: u32, cycle_row: usize) {
...
+/// Generates the expected and calculated constraint result on an input frame.
+fn get_bitwise_frame_result(frame: EvaluationFrame<Felt>, row: usize) -> [Felt; NUM_CONSTRAINTS] {
+    let cycle_row = get_test_periodic_values(row);
+    let mut result = [Felt::ZERO; NUM_CONSTRAINTS];
+    let _expected = result;
+
+    bitwise::enforce_constraints(&frame, &cycle_row, &mut result, Felt::ONE);
+
+    result
+}
```
Comment (grjte) [inline air/src/aux_table/bitwise_pow2/tests.rs:202]: The reference to "expected" in the comment can be removed and the following line can be deleted now that expected is not returned:

```
let _expected = result;
```

**Review by grjte — CHANGES_REQUESTED** [review]
Thank you! There's actually just one last small thing inline

**Review by grjte — APPROVED** [review]
Looks great, thank you! I'll go ahead and merge it. 

One thing to note - these tests are passing, but we will want a test that fails without the selector constraint changes that you're working on. This should be easy with the new test structure you've added - we can just add another selector test where using two different selectors generates the same output (forcing us to have a non-zero constraint evaluation somewhere else in the result array).

For example, we could test `XOR` and `OR` with 
```
let next_bitwise = vec![
        Felt::ZERO,
        Felt::ZERO,
        Felt::ZERO,
        Felt::ZERO,
        Felt::ONE,
        Felt::ONE,
        Felt::ONE,
        Felt::ONE,
    ];
```

This can be done in the next PR, since it'll be good to get this new test structure merged first.
